### PR TITLE
taking sub/superdiagonal argument checks too seriously

### PR DESCRIPTION
**Pull request by Sacha0**
Existing `tri(u|l)[!](A, k)` methods check that argument `k` refers to a valid (super)diagonal. In #14702, @tkelman and I noticed that existing checks allow superdiagonals `n` and `-m` for `m x n` matrices, which are out of bounds (given superdiagonals range from `n - 1` to `1 - m`). #14702 accordingly tightened these checks for new `SparseMatrixCSC` `tri(u|l)[!]` methods, but left the existing checks as they were.

This pull request's first commit tightens these checks elsewhere, introduces them where absent (`tri(u|l)(A::BitArray, k)`), makes the error thrown consistent, and revises a few tests that ran over out-of-bounds superdiagonals. (**Question**: The existing checks throw `ArgumentError("requested diagonal, $k$, out of bounds in matrix of size ($m, $n)")`s in most locations, but throw empty `BoundsError`s in others. Encountering the former more often, I standardized on it by default. Which is correct?)

This pull request's second commit accordingly tightens the tests for these checks.

Best!

cc @andreasnoack

### CODE REVIEW

**Comment by Sacha0** [comment]
Fixed and expanded failing `BitArray` tests.

**Comment by Sacha0** [comment]
AppVeyor timeout unrelated?

**Comment by andreasnoack** [comment]
Why do we throw here at all? Why not just let `k` be outside the matrix and then either leave or zero the matrix? I know this PR is just making things consistent but it's also a good oppertunity to discuss this. The behavior I describe is what Matlab does. I guess the usual trade offs apply: sometimes this check will reveal an error but sometimes it might make you algorithm a bit simpler.

**Comment by Sacha0** [comment]
> Why do we throw here at all? Why not just let k be outside the matrix and then either leave or zero the matrix? I know this PR is just making things consistent but it's also a good oppertunity to discuss this. The behavior I describe is what Matlab does. I guess the usual trade offs apply: sometimes this check will reveal an error but sometimes it might make you algorithm a bit simpler.

Cheers, worth discussing. I likewise see merits to both approaches and don't feel strongly one way or the other. Best!

**Comment by andreasnoack** [comment]
I think I'm in favor of removing the checks.

**Comment by Sacha0** [comment]
Thoughts @kshyatt @tkelman @KristofferC, other interested parties?

**Comment by tkelman** [comment]
We could have an unchecked, unexported inner function?

**Comment by Sacha0** [comment]
> We could have an unchecked, unexported inner function?

I would be equally happy: (1) removing argument-`k` checks in `tri(u|l)[!]` methods; (2) reimplementing `tri(u|l)[!]` methods as checked short children of corresponding `unchecked_tri(u|l)[!]` methods; or (3) revising argument-`k` checks in `tri(u|l)[!]` methods as in this PR. The question re. (2) is whether providing an argument-`k` checking escape hatch in `tri(u|l)[!]` methods is worth the additional complexity. Thoughts? Thanks and best!

**Comment by tkelman** [comment]
If you request some crazy way-out-of-bounds value for the number of off-diagonals you want, that's likely a bug and it would be a little unfortunate to have it pass through silently.

**Comment by Sacha0** [comment]
> If you request some crazy way-out-of-bounds value for the number of off-diagonals you want, that's likely a bug and it would be a little unfortunate to have it pass through silently.

Agreed. `@andreasnoack` points out a tradeoff in https://github.com/JuliaLang/julia/pull/16959#issuecomment-230965956. I lean slightly towards demanding correctness, i.e. options (2)/(3) over (1). Best!

**Comment by andreasnoack** [comment]
I think we should go for one or the other. It's not important enough to have both.

Minimizing the cases where we throw will give the smoother user experience and for this method, `k>n` can easily be justified. We can try to guess that the user meant something different but it's not wrong. It's also simpler to add extra checks than wrapping a throwing method in try/catch so overall I vote for removing the check. On the other hand, I don't feel strongly enough about this to block or delay any efforts to make this more consistent across methods.

**Comment by andreasnoack** [comment]
@Sacha0 What is your current thinking here? Do you still prefer throwing exceptions?

**Comment by Sacha0** [comment]
> What is your current thinking here?

Roughly the same. Above you mentioned

> `k > n` can easily be justified

Could you expand? Thanks!

**Comment by andreasnoack** [comment]
> Could you expand? Thanks!

The interpretation would be that elements above/below the `k`th diagonal will be zeroed whether or not the given matrix has a `k`th diagonal. If `k>n` then it is just all or none. It might be that you could write an algorithm that determines `n` and `k` independently and if it happens that `k>n` then it is totally fine that nothing in the matrix is changed. I guess this is the reason why Matlab allows this. On the other hand, this might have been a result of a programming error and you'd preferred the error but this is the usual tradeoff. Matlab often tries to minimize the cases where it throws in these situations and I'm sympathetic to that.

**Comment by Sacha0** [comment]
Could you point to a concrete example? I appreciate the abstract argument (thanks!) but have difficulty imagining a concrete case where throwing wouldn't be as good or better (to catch logical errors, or force clarification of code). Thanks!

**Comment by andreasnoack** [comment]
I agree that real examples are good when thinking about these things but I can't come up with a good one. It's mainly a recognition of the user experience associated with avoiding unnecessary exceptions. I believe it is part of Matlab's success but some people hate that behavior. However, feel free to decide which direction you want to follow. Getting it consistent would be great.

**Comment by Sacha0** [comment]
Cheers, in that case I still err to throwing exceptions --- would you be happy with throwing exceptions at this point? Thanks!

**Comment by Sacha0** [comment]
Rebased and revised a bit. What say you, CI?

**Comment by StefanKarpinski** [comment]
I actually think that the valid range of diagonals should be `[-size(A,1), size(A,2)]` with the extremes being degenerate cases where the diagonal has zero length. Having a negative length diagonal is legitimately invalid and seem likely to indicate a programming error, but what exactly is wrong with a zero-length diagonal given that everyone agrees that diagonal lengths from `minimum(size(A))` down to 1 are all valid – why is 0 suddenly wrong?

**Comment by tkelman** [comment]
This isn't asking for the diagonal though, it's asking for the array within it. So should there be two allowed boundary conditions both equivalent to getting the whole array back?

**Comment by StefanKarpinski** [comment]
There's only one way to get the whole array back / get all zeros back – asking for `tri[ul](A, -size(A,1))` and `tri[ul](A, size(A,2))`. My point is that the case where you're asking for everything above/below the zero-length diagonal is a valid corner case (literally), not an error. Asking for a diagonal beyond the corner of the matrix is most likely an error. I think we should broaden all triangular functions to allow asking for the zero-length diagonals in the corners of matrices, rather than making that an error everywhere.

**Comment by Sacha0** [comment]
These changes would be breaking. Perhaps should receive a 1.0 tag?

**Comment by Sacha0** [comment]
Partial highlights of a slack/#linalg conversation on this topic. (Unfortunately that conversation was lost to the sands of limited slack history and my recollection is incomplete. So the below surely misses some points and/or includes mistakes.)

Two broad proposals exist:

1) Enforce no constraints on sub-/super- diagonal index `k`
2) Enforce some constraints on sub-/super- diagonal index `k`

On the one hand, (1) has the advantage of never getting in the way. On the other hand, (2) has the advantage of potentially catching logical errors (assuming the constraints are adequately tight).

Regarding (2), the question is what precisely the constraints should be. On the one hand, constraints that are too loose mitigate (2)'s advantage. On the other hand, constraints that are too tight can, for example, impede generic programming by causing trouble in corner cases.

An m-by-n matrix has (non-empty) sub-/super- diagonals `-(m - 1)` through `n - 1`, necessitating the sub-/super- diagonal index constraint not be tighter than `-(m - 1) <= k <= n - 1`.

Should the constraint be looser?

Grepping base for uses of `tri[u|l][!]` and excluding those in definitions of `tri[u|l][!]` or in arbitrary examples, one (strictly?) finds uses in construction and/or manipulation of structured/special matrices. (Basically in constructing classical banded splittings of matrices.) Examples:

a) Constructon/manipulation of [unit] triangular matrices, with sub-/super- diagonal index either one, zero, or negative one. Though somewhat strange, a one-by-one triangular matrix could form a sensical corner case or base case in a recursion. So arguably `tri[u|l][!](A, [-]1)` should work for one-by-one `A`, suggesting the sub-/super- diagonal index constraint not be tighter than `-m <= k <= n`.

Essentially identical statements could be made for trapezoidal matrices, though trapezoidal matrices do not exist in base.

Similar statements could be made for Hessenberg matrices (which do exist in base), but with sub-/super- diagonal index either two or negative two. But a lower(upper) Hessenberg matrix with fewer than two columns(rows) makes little sense. So perhaps constructing/manipulating Hessenberg matrices also requires only `-m <= k <= n` at tightest?

b) Construction/manipulation of diagonal matrices, with sub-/super- diagonal index either one or negative one. A one-by-one diagonal matrix certainly makes sense. So `tri[u|l][!](A, [-]1)` should work for one-by-one `A`, also necessitating the sub-/super- diagonal index constraint not be tigher than `-m <= k <= n`.

c) Construction/manipulation of upper(lower) bidiagonal matrices, with sub-/super- diagonal index either two(one) or negative one(two). But an upper(lower) bidiagonal matrix with fewer than two columns(rows) makes little sense. So perhaps constructing/manipulating bidiagonal matrices also requires only `-m <= k <= n` at tightest?

d) Construction/manipulation of tridiagonal matrices, with sub-/super- diagonal index either two or negative two. But a tridiagonal matrix with fewer than two columns and two rows makes little sense. So perhaps constructing/manipulating tridiagonal matrices also requires only `-m <= k <= n` at tightest?

The generalization of these examples is band matrices(/splittings) in general. Construction/manipulation of a band matrix(/splitting) with upper bandwidth `u` and lower bandwidth `l` should require at broadest `tri[u|l][!](A, [(u+1)|(l-1)])`, requiring only `-m <= k <= n` at tightest?

But IIRC, the slack/#linalg conversation came up with asymmetric bounds for sub- and super- diagonal index checks, resulting in near-consensus of enforcing no constraints. Is the above somehow in error, or missing use cases? The primary participants in that conversation were @StefanKarpinski, @andreasnoack, and @mbauman. Do any of you remember where the asymmetric bounds came from? (Considering only one type of Hessenberg matrix?) And/or other details missing above? Thanks and best!

**Comment by andreasnoack** [comment]
Thanks for the summary. I don't remember where asymmetric bounds come from.

**Comment by StefanKarpinski** [comment]
The key problem is that you often see code that decomposes matrices into upper + diagonal + lower:
```jl
A = tril(A, k-1) + diag(A, k) + triu(A, k+1)
```
Since `triu` and `tril` include the diagonal by design, the diagonal offset value needs to be one lower and one higher for them compared to `diag`. The value of `k` above naturally ranges from `k = -size(A,1):size(A,2)` – including the "zero-length diagonals" in the corners of a matrix. These are the cases where the upper or lower component get all of the values in the decomposition. This case really happens and is very annoying to write if one doesn't allow diagonal values in the corners. Moreover, even if one thinks that case is "bad" or "wrong" or should be expressed with some awkward conditional expression instead, there's also the degenerate 0x0 matrix case, where the the only possible value `k` can take is 0. So unless one wants to *always* handle zero dimensions specially when taking diagonals, one needs to allow at least the range `-size(A,1):size(A,2)` for the `diag` function.

Since `diag` is so often used together with `triu` and/or `tril` and they need ±1 offsets in the decomposition, we _also_ need to allow some values of `k` outside of `-size(A,1):size(A,2)` for them. Specifically, we need to allow at least `-size(A,1)-1:size(A,2)-1` for `tril` and `-size(A,1)+1:size(A,2)+1` for `triu`. So we are faced with three choices:

1. Each of `diag`, `tril` and `triu` has their own, idiosyncratic range but precise range of allowable diagonal values:
  * `diag`: `-size(A,1):size(A,2)`
  * `tril`: `-size(A,1)-1:size(A,2)-1`
  * `triu`: `-size(A,1)+1:size(A,2)+1`
This option is precise but inconsistent. I suspect that it will be pretty hard to explain to people why all the functions that take diagonal offsets have different ranges of allowable offsets.

2. Have a single common range of allowable diagonals which includes some corner cases for each function that don't make sense for that function: `-size(A,1)-1:size(A,2)+1`. This avoids the inconsistency of the first solution at the cost of failing to catch precisely the off-by-one errors that one would hope to catch by throwing errors for invalid diagonal values. Now you need an off-by-two error most of the time in order to catch mistakes.

3. Don't bounds check diagonal offsets at all and just apply the obvious conditional zero/value formula for each function comparing the sum of indices to the offset.

**Comment by StefanKarpinski** [comment]
In other words, for an arbitrary matrix `A`, you want to be able to write this and have it work:
```jl
for k = -size(A,1):size(A,2)
    @assert A == tril(A,k-1) + diag(A,k) + triu(A,k+1)
end
```
Currently, this errors at the corners:
```
ERROR: ArgumentError: requested diagonal, -4, out of bounds in matrix of size (3,4)
```
So really, the correct subject of a PR for this should be to broaden the allowable range of diagonal offsets so that we can run that code without an error.

**Comment by Sacha0** [comment]
Much thanks for filling in the missing points! :) Out of curiosity, where does the splitting `tril(A, k-1) + diag(A, k) + triu(A, k+1)` with`k != 0` occur in practice? Thanks!

**Comment by StefanKarpinski** [comment]
I don't recall exactly, but we had looked for Julia repos using `tril` and `triu` and a lot of them used this pattern to decompose matrices. And a lot of them were actually broken in the corner cases!

**Comment by StefanKarpinski** [comment]
My inclination is either option 1 or 3 – precise (inconsistent) bounds or no bounds for diagonals. The second option seems like a worst of both worlds choice to me. The conservative choice is 1 for now and see if we get complaints or problem due to the inconsistent bounds for the three functions. If not, then it's fine; if so, we can always loosen the bounds without breaking working code.

**Comment by StefanKarpinski** [comment]
Bump, @Sacha0 – any way you could dust this off with option 1?

**Comment by Sacha0** [comment]
Sounds good! :)

**Comment by Sacha0** [comment]
Rewritten for option one (less the forthcoming changes to `diag`). Passed CI prior to a conflict rebase. Best!

**Comment by Sacha0** [comment]
Added a commit that makes `diag` consistent with option one as well. Best!

**Review by StefanKarpinski — APPROVED** [review]

File: base/linalg/bidiag.jl [hunk 1]
```diff
@@ -271,10 +273,11 @@ function diag(M::Bidiagonal{T}, n::Integer=0) where T
         return M.uplo == 'U' ? M.ev : zeros(T, size(M,1)-1)
     elseif n == -1
         return M.uplo == 'L' ? M.ev : zeros(T, size(M,1)-1)
-    elseif -size(M,1) < n < size(M,1)
+    elseif -size(M,1) <= n <= size(M,1)
```
Comment (jebej) [inline base/linalg/bidiag.jl]: Is the upper bound meant to be size(M,2) here?

Comment (jebej) [reply base/linalg/bidiag.jl]: Or possibly size(M,1) instead of 2 in the exception below

Comment (andreasnoack) [reply base/linalg/bidiag.jl]: `M` is square so it doesn't matter

**Comment by Sacha0** [comment]
Thanks all! :)

**Pull request merged**